File: llamafile/server/www/highlight_markdown.js
// Copyright 2024 Mozilla Foundation
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

class HighlightMarkdown extends Highlighter {

  static NORMAL = 0;
  static TICK = 1;
  static TICK_TICK = 2;
  static CODE = 4;
  static CODE_TICK = 5;
  static CODE_TICK_TICK = 6;
  static STAR = 7;
  static STRONG = 8;
  static STRONG_BACKSLASH = 9;
  static STRONG_STAR = 10;
  static BACKSLASH = 11;
  static INCODE = 12;
  static INCODE2 = 13;
  static INCODE2_TICK = 14;
  static INCODE2_TICK2 = 15;
  static EMPHASIS = 16;
  static EMPHASIS_BACKSLASH = 17;

  constructor(delegate) {
    super(delegate);
    this.bol = true;
    this.tail = false;
  }

  feed(input) {
    for (let i = 0; i < input.length; i += this.delta) {
      this.delta = 1;
      let c = input[i];
      if (c == '\u0000')
        c = '\ufffd';
      switch (this.state) {

      case HighlightMarkdown.NORMAL:
        if (c == '`') {
          this.state = HighlightMarkdown.TICK;
          break;
        } else if (c == '*') {
          this.state = HighlightMarkdown.STAR;
          break;
        } else if (c == '\\') {
          // handle \*\*not bold\*\* etc.
          this.state = HighlightMarkdown.BACKSLASH;
          this.append('\\');
          this.bol = false;
        } else {
          this.append(c);
        }
        if (c == '\n') {
          this.bol = true;
          this.tail = false;
        } else {
          this.tail = true;
          if (!isblank(c))
            this.bol = false;
        }
        break;

      case HighlightMarkdown.BACKSLASH:
        this.append(c);
        this.state = HighlightMarkdown.NORMAL;
        break;

      case HighlightMarkdown.STAR:
        if (c == '*') {
          // handle **strong** text
          this.state = HighlightMarkdown.STRONG;
          this.push("STRONG", "");
          this.append("**");
        } else if (this.bol && isblank(c)) {
          this.append('*');
          this.append(c);
          this.state = HighlightMarkdown.NORMAL;
        } else {
          // handle *emphasized* text
          // inverted because \e[3m has a poorly supported western bias
          this.append('*');
          this.push("EM", "");
          this.append(c);
          this.state = HighlightMarkdown.EMPHASIS;
          if (c == '\\')
            this.state = HighlightMarkdown.EMPHASIS_BACKSLASH;
        }
        this.bol = false;
        break;

      case HighlightMarkdown.EMPHASIS:
        // this is for *emphasized* text
        if (c == '*') {
          this.state = HighlightMarkdown.NORMAL;
          this.pop();
          this.append('*');
        } else if (c == '\\') {
          this.state = HighlightMarkdown.EMPHASIS_BACKSLASH;
          this.append('\\');
        } else {
          this.append(c);
        }
        break;

      case HighlightMarkdown.EMPHASIS_BACKSLASH:
        // so we can say *unbroken \* italic* and have it work
        this.append(c);
        this.state = HighlightMarkdown.EMPHASIS;
        break;

      case HighlightMarkdown.STRONG:
        this.append(c);
        if (c == '*') {
          this.state = HighlightMarkdown.STRONG_STAR;
        } else if (c == '\\') {
          this.state = HighlightMarkdown.STRONG_BACKSLASH;
        }
        break;

      case HighlightMarkdown.STRONG_BACKSLASH:
        // so we can say **unbroken \*\* bold** and have it work
        this.append(c);
        this.state = HighlightMarkdown.STRONG;
        break;

      case HighlightMarkdown.STRONG_STAR:
        this.append(c);
        if (c == '*' || // handle **bold** ending
            (c == '\n' && !this.tail)) { // handle *** line break
          this.state = HighlightMarkdown.NORMAL;
          this.pop();
        } else if (c == '\\') {
          this.state = HighlightMarkdown.STRONG_BACKSLASH;
        } else {
          this.state = HighlightMarkdown.STRONG;
        }
        break;

      case HighlightMarkdown.TICK:
        if (c == '`') {
          if (this.bol) {
            this.state = HighlightMarkdown.TICK_TICK;
          } else {
            this.push("SPAN", "incode");
            this.append("``");
            this.state = HighlightMarkdown.INCODE2;
          }
        } else {
          this.push("SPAN", "incode");
          this.append('`');
          this.append(c);
          this.state = HighlightMarkdown.INCODE;
        }
        this.bol = false;
        break;

      case HighlightMarkdown.INCODE:
        // this is for `inline code` like that
        // no backslash escapes are supported here
        this.append(c);
        if (c == '`') {
          this.pop();
          this.state = HighlightMarkdown.NORMAL;
        }
        break;

      case HighlightMarkdown.INCODE2:
        // this is for ``inline ` code`` like that
        // it lets you put backtick inside the code
        this.append(c);
        if (c == '`') {
          this.state = HighlightMarkdown.INCODE2_TICK;
        }
        break;

      case HighlightMarkdown.INCODE2_TICK:
        this.append(c);
        if (c == '`') {
          this.state = HighlightMarkdown.INCODE2_TICK2;
        } else {
          this.state = HighlightMarkdown.INCODE2;
        }
        break;

      case HighlightMarkdown.INCODE2_TICK2:
        if (c == '`') {
          this.append('`');
        } else {
          this.pop();
          this.epsilon(HighlightMarkdown.NORMAL);
        }
        break;

      case HighlightMarkdown.TICK_TICK:
        if (c == '`') {
          this.state = HighlightMarkdown.CODE;
          this.append("```");
        } else {
          this.push("SPAN", "incode");
          this.append("``");
          this.append(c);
          this.state = HighlightMarkdown.INCODE2;
        }
        break;

      case HighlightMarkdown.CODE:
        if (c == '`') {
          this.state = HighlightMarkdown.CODE_TICK;
        } else {
          this.append(c);
        }
        break;

      case HighlightMarkdown.CODE_TICK:
        if (c == '`') {
          this.state = HighlightMarkdown.CODE_TICK_TICK;
        } else {
          this.append('`');
          this.epsilon(HighlightMarkdown.CODE);
        }
        break;

      case HighlightMarkdown.CODE_TICK_TICK:
        if (c == '`') {
          this.state = HighlightMarkdown.NORMAL;
          this.append("```");
        } else {
          this.append("``");
          this.epsilon(HighlightMarkdown.CODE);
        }
        break;

      default:
        throw new Error('Invalid state');
      }
    }
  }

  flush() {
    switch (this.state) {
    case HighlightMarkdown.STAR:
      this.append('*');
      break;
    case HighlightMarkdown.TICK:
    case HighlightMarkdown.CODE_TICK:
      this.append('`');
      break;
    case HighlightMarkdown.TICK_TICK:
    case HighlightMarkdown.CODE_TICK_TICK:
      this.append("``");
      break;
    case HighlightMarkdown.INCODE:
    case HighlightMarkdown.INCODE2:
    case HighlightMarkdown.INCODE2_TICK:
    case HighlightMarkdown.INCODE2_TICK2:
    case HighlightMarkdown.STRONG:
    case HighlightMarkdown.STRONG_BACKSLASH:
    case HighlightMarkdown.STRONG_STAR:
    case HighlightMarkdown.EMPHASIS:
    case HighlightMarkdown.EMPHASIS_BACKSLASH:
      this.pop();
      break;
    default:
      break;
    }
    this.state = HighlightMarkdown.NORMAL;
    this.bol = true;
    this.tail = false;
    this.delegate.flush();
  }
}

Highlighter.REGISTRY['markdown'] = HighlightMarkdown;
Highlighter.REGISTRY['md'] = HighlightMarkdown;

File: llamafile/highlight/highlight_markdown.cpp
// -*- mode:c++;indent-tabs-mode:nil;c-basic-offset:4;coding:utf-8 -*-
// vi: set et ft=cpp ts=4 sts=4 sw=4 fenc=utf-8 :vi
//
// Copyright 2024 Mozilla Foundation
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#include "highlight.h"
#include "util.h"
#include <cosmo.h>

enum
{
    NORMAL,
    TICK,
    TICK_TICK,
    LANG,
    CODE,
    CODE_TICK,
    CODE_TICK_TICK,
    STAR,
    STRONG,
    STRONG_BACKSLASH,
    STRONG_STAR,
    BACKSLASH,
    INCODE,
    INCODE2,
    INCODE2_TICK,
    INCODE2_TICK2,
    EMPHASIS,
    EMPHASIS_BACKSLASH,
};

HighlightMarkdown::HighlightMarkdown()
{
}

HighlightMarkdown::~HighlightMarkdown()
{
}

void
HighlightMarkdown::feed(std::string* r, std::string_view input)
{
    for (size_t i = 0; i < input.size(); ++i) {
        wchar_t c;
        int b = input[i] & 255;
        if (!u_) {
            if (b < 0300) {
                c = b;
            } else {
                c_ = ThomPikeByte(b);
                u_ = ThomPikeLen(b) - 1;
                continue;
            }
        } else if (ThomPikeCont(b)) {
            c = c_ = ThomPikeMerge(c_, b);
            if (--u_)
                continue;
        } else {
            u_ = 0;
            c = b;
        }
        if (c == 0xFEFF)
            continue; // utf-8 bom
        if (!c)
            c = 0xfffd; // replacement character
        switch (t_) {

        Normal:
        case NORMAL:
            if (c == '`') {
                t_ = TICK;
                break;
            } else if (c == '*') {
                t_ = STAR;
                break;
            } else if (c == '\\') {
                // handle \*\*not bold\*\* etc.
                t_ = BACKSLASH;
                *r += '\\';
                bol_ = false;
            } else {
                lf::append_wchar(r, c);
            }
            if (c == '\n') {
                bol_ = true;
                tail_ = false;
            } else {
                tail_ = true;
                if (!isblank(c))
                    bol_ = false;
            }
            break;

        case BACKSLASH:
            lf::append_wchar(r, c);
            t_ = NORMAL;
            break;

        case STAR:
            if (c == '*') {
                // handle **strong** text
                t_ = STRONG;
                *r += HI_BOLD;
                *r += "**";
            } else if (bol_ && isblank(c)) {
                *r += '*';
                lf::append_wchar(r, c);
                t_ = NORMAL;
            } else {
                // handle *emphasized* text
                // inverted because \e[3m has a poorly supported western
                // bias
                *r += '*';
                *r += HI_ITALIC;
                lf::append_wchar(r, c);
                t_ = EMPHASIS;
                if (c == '\\')
                    t_ = EMPHASIS_BACKSLASH;
            }
            bol_ = false;
            break;

        case EMPHASIS:
            // this is for *emphasized* text
            if (c == '*') {
                t_ = NORMAL;
                *r += HI_RESET;
                *r += '*';
            } else if (c == '\\') {
                t_ = EMPHASIS_BACKSLASH;
                *r += '\\';
            } else {
                lf::append_wchar(r, c);
            }
            break;

        case EMPHASIS_BACKSLASH:
            // so we can say *unbroken \* italic* and have it work
            lf::append_wchar(r, c);
            t_ = EMPHASIS;
            break;

        case STRONG:
            lf::append_wchar(r, c);
            if (c == '*') {
                t_ = STRONG_STAR;
            } else if (c == '\\') {
                t_ = STRONG_BACKSLASH;
            }
            break;

        case STRONG_BACKSLASH:
            // so we can say **unbroken \*\* bold** and have it work
            lf::append_wchar(r, c);
            t_ = STRONG;
            break;

        case STRONG_STAR:
            lf::append_wchar(r, c);
            if (c == '*' || // handle **bold** ending
                (c == '\n' && !tail_)) { // handle *** line break
                t_ = NORMAL;
                *r += HI_RESET;
            } else if (c == '\\') {
                t_ = STRONG_BACKSLASH;
            } else {
                t_ = STRONG;
            }
            break;

        case TICK:
            if (c == '`') {
                if (bol_) {
                    t_ = TICK_TICK;
                } else {
                    *r += HI_INCODE;
                    *r += "``";
                    t_ = INCODE2;
                }
            } else {
                *r += HI_INCODE;
                *r += '`';
                lf::append_wchar(r, c);
                t_ = INCODE;
            }
            bol_ = false;
            break;

        case INCODE:
            // this is for `inline code` like that
            // no backslash escapes are supported here
            lf::append_wchar(r, c);
            if (c == '`') {
                *r += HI_RESET;
                t_ = NORMAL;
            }
            break;

        case INCODE2:
            // this is for ``inline ` code`` like that
            // it lets you put backtick inside the code
            lf::append_wchar(r, c);
            if (c == '`') {
                t_ = INCODE2_TICK;
            }
            break;

        case INCODE2_TICK:
            lf::append_wchar(r, c);
            if (c == '`') {
                t_ = INCODE2_TICK2;
            } else {
                t_ = INCODE2;
            }
            break;

        case INCODE2_TICK2:
            if (c == '`') {
                *r += '`';
            } else {
                *r += HI_RESET;
                t_ = NORMAL;
                goto Normal;
            }
            break;

        case TICK_TICK:
            if (c == '`') {
                t_ = LANG;
                *r += "```";
            } else {
                *r += HI_INCODE;
                *r += "``";
                lf::append_wchar(r, c);
                t_ = INCODE2;
            }
            break;

        case LANG:
            if (!isascii(c) || !isspace(c)) {
                lf::append_wchar(r, c);
                lang_ += std::tolower(c);
            } else {
                if (!(highlighter_ = Highlight::create(lang_)))
                    highlighter_ = new HighlightTxt;
                lang_.clear();
                t_ = CODE;
                goto Code;
            }
            break;

        Code:
        case CODE:
            if (c == '`') {
                t_ = CODE_TICK;
            } else {
                char cs[8];
                uint64_t w = tpenc(c);
                WRITE64LE(cs, w);
                highlighter_->feed(r, cs);
            }
            break;

        case CODE_TICK:
            if (c == '`') {
                t_ = CODE_TICK_TICK;
            } else {
                char cs[9];
                uint64_t w = tpenc(c);
                cs[0] = '`';
                WRITE64LE(cs + 1, w);
                highlighter_->feed(r, cs);
                t_ = CODE;
            }
            break;

        case CODE_TICK_TICK:
            if (c == '`') {
                t_ = NORMAL;
                highlighter_->flush(r);
                delete highlighter_;
                highlighter_ = nullptr;
                *r += "```";
            } else {
                char cs[10];
                uint64_t w = tpenc(c);
                cs[0] = '`';
                cs[1] = '`';
                WRITE64LE(cs + 2, w);
                highlighter_->feed(r, cs);
                t_ = CODE;
            }
            break;

        default:
            __builtin_unreachable();
        }
    }
}

void
HighlightMarkdown::flush(std::string* r)
{
    switch (t_) {
    case LANG:
        lang_.clear();
        break;
    case STAR:
        *r += '*';
        break;
    case TICK:
        *r += '`';
        break;
    case TICK_TICK:
        *r += "``";
        break;
    case INCODE:
    case INCODE2:
    case INCODE2_TICK:
    case INCODE2_TICK2:
    case STRONG:
    case STRONG_BACKSLASH:
    case STRONG_STAR:
    case EMPHASIS:
    case EMPHASIS_BACKSLASH:
        *r += HI_RESET;
        break;
    case CODE:
        highlighter_->flush(r);
        delete highlighter_;
        highlighter_ = nullptr;
        break;
    case CODE_TICK:
        highlighter_->flush(r);
        delete highlighter_;
        highlighter_ = nullptr;
        *r += '`';
        break;
    case CODE_TICK_TICK:
        highlighter_->flush(r);
        delete highlighter_;
        highlighter_ = nullptr;
        *r += "``";
        break;
    default:
        break;
    }
    c_ = 0;
    u_ = 0;
    t_ = NORMAL;
    bol_ = true;
    tail_ = false;
}
